src/Raven.Server/Documents/Sharding/Handlers/ShardedIncomingReplicationHandler.cs

```diff
@@ -140,8 +166,101 @@ internal async Task<(string MergedChangeVector, long LastEtag)> GetInitialHandsh
             return (mergedChangeVector, lastAcceptedEtag);
         }
 
-        protected override void InvokeOnDocumentsReceived()
+        private async Task<(string MergedChangeVector, long LastEtag)> GetInitialHandshakeResponseForShardedSourceAsync()
         {
+            var shardedExternalReplicationStates = ShardReplicationLoader.GetShardedExternalReplicationStates(_parent.Server, _parent.DatabaseName,
+                    _sourceShardedDatabaseName, _sourceShardedDatabaseId);
+
+            if (shardedExternalReplicationStates == null ||
+                shardedExternalReplicationStates.ReplicationStates.TryGetValue(ConnectionInfo.SourceDatabaseName, out var stateForSingleSource) == false ||
+                string.IsNullOrEmpty(stateForSingleSource.LastSourceChangeVector))
+                return (null, 0);
+
+            var lastAcceptedEtag = stateForSingleSource.LastSourceEtag;
+            List<string> handlersChangeVector = null;
+            foreach (var (destinationDatabaseName, destinationState) in stateForSingleSource.DestinationStates)
+            {
+                foreach (var (_, handler) in _handlers)
+                {
+                    if (string.Equals(handler.DatabaseName, destinationDatabaseName, StringComparison.OrdinalIgnoreCase) == false)
+                        continue;
+
+                    var (acceptedChangeVector, acceptedEtag) = await handler.GetFirstChangeVectorFromShardAsync();
+                    if (string.IsNullOrEmpty(acceptedChangeVector))
+                        return (null, 0);
+
+                    handlersChangeVector ??= new List<string> { stateForSingleSource.LastSourceChangeVector };
+                    handlersChangeVector.Add(destinationState.DestinationChangeVector);
+                    handlersChangeVector.Add(acceptedChangeVector);
+                    lastAcceptedEtag = Math.Min(acceptedEtag, lastAcceptedEtag);
+                }
+            }
+
+            var mergedChangeVector = ChangeVectorUtils.MergeVectorsDown(handlersChangeVector);
+            return (mergedChangeVector, lastAcceptedEtag);
+        }
+
+        protected override void OnDocumentsReceived()
+        {
+            if (_shardedSource == false)
+                return;
+
+            var states = ShardReplicationLoader.GetShardedExternalReplicationStates(_parent.Server, _parent.DatabaseName, _sourceShardedDatabaseName, _sourceShardedDatabaseId);
+
+            states ??= new ShardedExternalReplicationState
```

Comment:
the update of the state should not be done here, but in the cluster state machine.
You should sent what should be updated.
This way we can easier handle concurrency if required.

Reply:
done